libs/core/langchain_core/runnables/factory.py

```diff
@@ -0,0 +1,189 @@
+from abc import abstractmethod
+from typing import Any, AsyncIterator, Iterator, List, Optional, Sequence, Type, Union
+
+from langchain_core.pydantic_v1 import BaseModel, PrivateAttr
+from langchain_core.runnables import (
+    Runnable,
+    RunnableConfig,
+    RunnableSerializable,
+)
+from langchain_core.runnables.graph import Graph
+from langchain_core.runnables.utils import ConfigurableFieldSpec, Input, Output
+
+
+class RunnableFactory(RunnableSerializable[Input, Output]):
+    """Runnable that creates another runnable at runtime.
+
+    The runnable allows creating LCEL chains using object-oriented programming.
+    It provides a factory method :meth:`_create_runnable` to build another runnable
+    at runtime.
+
+    Features:
+
+    * Loadable parameters and validation thanks to ``pydantic``.
+    * Inheritance support and all other class features.
+
+    Example usage:
+
+    .. code-block:: python
+
+        from cogent_core.utils.runnable import RunnableFactory
+        from langchain_core.output_parsers import StrOutputParser
+        from langchain_core.prompts import ChatPromptTemplate
+        from langchain_core.runnables import Runnable
+        from langchain_openai import ChatOpenAI
+        from langchain_core.pydantic_v1 import validator
+
+        class MyRunnable(RunnableFactory):
+            role: str = "You are helpful AI assistant."
+            prompt: str = "Tell me a joke about {topic}"
+            model: str = "gpt-3.5-turbo"
+
+            @validator("prompt")
+            def validate_prompt(cls, v):
+                if "joke" not in v:
+                    raise ValueError("You should only ask for jokes.")
+                return v
+
+            def _create_runnable(self) -> Runnable:
+                prompt = ChatPromptTemplate.from_messages(
+                    [("system", self.role), ("human", self.prompt)]
+                )
+                llm = ChatOpenAI(model_name=self.model)
+
+                chain = prompt | llm | StrOutputParser()
+                return chain
+
+        # Create a chain with defaults.
+
+        chain = MyRunnable()
+        output = chain.invoke({"topic": "bear"})
+        print(output)
+
+        # Create a chain with custom parameters.
+
+        chain = MyRunnable(model="gpt-4-1106-preview")
+        output = chain.invoke({"topic": "bear"})
+        print(output)
+
+        # Create a chain from external "source".
+
+        external_parameters = {
+            "model": "gpt-4-1106-preview",
+            "prompt": "Tell me two jokes about {topic}",
+        }
+
+        chain = MyRunnable.parse_obj(external_parameters)
+        output = chain.invoke({"topic": "bear"})
+        print(output)
+    """
+
+    _runnable: Runnable = PrivateAttr()
+
+    def __init__(self, **kwargs: Any) -> None:
+        super().__init__(**kwargs)
+        self._runnable = self._create_runnable()
```

Comment:
if the runnable is created on init of the factory then it's not really a factory is it? you could've just initialized the underlying runnable, no?

Reply:
That's right. The name is not very clear. Maybe `RunnableWrapper` or `RunnablePimpl` or smth else is better. One of the goals is to redirect API calls to the underlying runnable created at runtime. Let me know if it makes sense to put this whole idea in the core.